src/main/java/org/jenkinsci/plugins/ghprb/GhprbRootAction.java

```diff
@@ -45,33 +47,32 @@ public String getUrlName() {
 
     public void doIndex(StaplerRequest req, StaplerResponse resp) {
         String event = req.getHeader("X-GitHub-Event");
+        String signature = req.getHeader("X-Hub-Signature");
         String type = req.getContentType();
         String payload = null;
+        String body = null;
 
         if ("application/json".equals(type)) {
-            BufferedReader br = null;
-            try {
-                br = req.getReader();
-                payload = IOUtils.toString(req.getReader());
-            } catch (IOException e) {
+            body = extractRequestBody(req);
+            if (body == null) {
                 logger.log(Level.SEVERE, "Can't get request body for application/json.");
                 return;
-            } finally {
-                IOUtils.closeQuietly(br);
             }
+            payload = body;
         } else if ("application/x-www-form-urlencoded".equals(type)) {
-            payload = req.getParameter("payload");
-            if (payload == null) {
+            body = extractRequestBody(req);
+            if (body == null || body.length() <= 8) {
                 logger.log(Level.SEVERE, "Request doesn't contain payload. You're sending url encoded request, so you should pass github payload through 'payload' request parameter");
                 return;
             }
+            try {
+                payload = URLDecoder.decode(body.substring(8), req.getCharacterEncoding());
+            } catch (UnsupportedEncodingException e) {
+                logger.log(Level.SEVERE, "Error while trying to encode the payload");
+                return;
+            }
         }
 
-        if (payload == null) {
```

Comment:
Because the payload is no longer directly taken "as is". Instead, the full body is taken. If the POST is a json then payload == actual json == body, but if it is an encoded URL then the body is something like:
payload=<encoded payload here>
The reason we discard the payload and we take the body is because the signature has been made by github using the full body, so now I extract the body, and from there I extract the payload.